ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -450,35 +459,48 @@ func getReplicaIndex(sliceToWorkerIDs map[slice][]int, clusterName string, group
 }
 
 // getNextWorkerID returns the next lowest TPU_WORKER_ID in the Pod Slice
-func getNextWorkerID(sliceToWorkerIDs map[slice][]int, podSlice slice, namespace string, replicaIndex int) int {
+func getNextWorkerID(sliceToWorkerIDs map[slice][]int, podSlice slice, namespace string, replicaIndex int) (int, error) {
 	tpuWorkerID := 0 // defaults to 0 (first Pod in slice)
 	if len(sliceToWorkerIDs) == 0 || len(sliceToWorkerIDs[podSlice]) == 0 {
-		return tpuWorkerID
+		return tpuWorkerID, nil
 	}
 	sort.Ints(sliceToWorkerIDs[podSlice])
 	// iterate through existing workers and get the next lowest, unused ID
-	for _, workerID := range sliceToWorkerIDs[podSlice] {
+	lastID := 0
+	for index, workerID := range sliceToWorkerIDs[podSlice] {
+		// check for incorrect assignment of IDs
+		if index == 0 {
+			lastID = workerID
+		} else if workerID == lastID {
+			return 0, errors.New("Identical TPU_WORKER_ID assigned to multiple TPU workers in slice")
+		}
+		// get the next lowest, valid TPU_WORKER_ID
 		if workerID != tpuWorkerID {
 			break
 		}
+		lastID = workerID
 		tpuWorkerID++
 	}
-
-	klog.V(1).InfoS("getNextWorkerID", "RayCluster", namespace+"/"+podSlice.clusterName, "Worker Group", podSlice.groupName, "TPU_WORKER_ID", tpuWorkerID)
-	return tpuWorkerID
+	klog.V(1).InfoS("getNextWorkerID", "RayCluster", namespace+"/"+podSlice.clusterName, "Worker Group", podSlice.groupName, "replicaIndex", replicaIndex, "TPU_WORKER_ID", tpuWorkerID)
+	return tpuWorkerID, nil
 }
 
 // getSliceToWorkerIDs returns a mapping representing the current RayCluster state of TPU pods using a PodLister
-func getSliceToWorkerIDs(podsInGroup []*corev1.Pod, clusterName string, groupName string, namespace string, numOfHosts int32) (map[slice][]int, error) {
+func (t *TPUWebhookServer) getSliceToWorkerIDs(clusterName string, groupName string, namespace string, numOfHosts int32) (map[slice][]int, error) {
 	sliceToWorkerIDs := make(map[slice][]int)
 
+	podsInGroup, err := t.podLister.Pods(namespace).List(labels.SelectorFromSet(labels.Set{"ray.io/group": groupName}))
```

Comment:
should this also have ray.io/cluster=<cluster-name> in the selector? Otherwise clusters with similar group names would end up in `podsInGroup`

Reply:
Ah, I see later in the code we filter out pods not in the same cluster

Reply:
I went ahead and added `ray.io/cluster` to the selector in [d1a44cc](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/859/commits/d1a44cc66cf31b3abf91c72498369d3c098ce572) so that we get a shorter list of Pods, but yeah we were checking for group, cluster, and namespace when iterating through the Pods.